Provider/src/main/java/com/spotify/confidence/AndroidLifecycleEventProducer.kt

```diff
@@ -0,0 +1,237 @@
+
+    override fun onActivityDestroyed(activity: Activity) {
+        activity.trackActivity("activity-destroyed")
+    }
+
+    private fun trackApplicationLifecycleEvents() {
+        // Get the current version.
+        val packageInfo = packageInfo
+        val currentVersion = ConfidenceValue.String(packageInfo?.versionName ?: "")
+        val currentBuild = ConfidenceValue.String(packageInfo?.getVersionCode().toString() ?: "")
+
+        // Get the previous recorded version.
+        val previousVersion = sharedPreferences
+            .getString(APP_VERSION, null)
+            ?.let(ConfidenceValue::String)
+
+        val previousBuild: ConfidenceValue.String? = sharedPreferences
+            .getString(APP_BUILD, null)
+            ?.let(ConfidenceValue::String)
+
+        val legacyPreviousBuild = sharedPreferences.getString(LEGACY_APP_BUILD, null)
+
+        // Check and track Application Installed or Application Updated.
+        if (previousBuild == null && legacyPreviousBuild == null) {
+            val message = mapOf("version" to currentVersion, "build" to currentBuild)
+            coroutineScope.launch { eventsFlow.emit(Event(APP_INSTALLED_EVENT, message)) }
+        } else if (currentBuild != previousBuild) {
+            val message = mapOf(
+                "version" to currentVersion,
+                "build" to currentBuild,
+                "previous_version" to (previousVersion ?: ConfidenceValue.String("")),
+                "previous_build" to (previousBuild ?: ConfidenceValue.String(""))
+            )
+            coroutineScope.launch { eventsFlow.emit(Event(APP_UPDATED_EVENT, message)) }
+        }
+
+        coroutineScope.launch {
+            sharedPreferences.edit().putString(APP_VERSION, currentVersion.string).apply()
+            sharedPreferences.edit().putString(APP_BUILD, currentBuild.string).apply()
+        }
+
+        coroutineScope.launch {
+            val message = mapOf("version" to currentVersion, "build" to currentBuild)
+            eventsFlow.emit(Event(APP_LAUNCHED_EVENT, message))
+        }
+    }
+
+    override fun events(): Flow<Event> = eventsFlow
+
+    override fun contextChanges(): Flow<Map<String, ConfidenceValue>> = contextFlow
+    override fun stop() {
+        application.unregisterActivityLifecycleCallbacks(this)
+        coroutineScope.launch(Dispatchers.Main) {
+            lifecycle.removeObserver(this@AndroidLifecycleEventProducer)
+        }
+    }
+
+    companion object {
+        // SharedPreferences keys
+        private const val APP_VERSION = "APP_VERSION"
+        private const val APP_BUILD = "APP_BUILD"
+        private const val LEGACY_APP_BUILD = "LEGACY_APP_BUILD"
+
+        // Context keys
+        private const val IS_FOREGROUND_KEY = "is_foreground"
+
+        // Event keys
+        private const val APP_INSTALLED_EVENT = "app-installed"
+        private const val APP_UPDATED_EVENT = "app-updated"
```

Comment:
While manual testing, I noticed these are fired more than once in a single session